File: tools/imageDefogging/src/autoWhiteBalance.cpp
/*
 * 文件名: autoWhiteBalance.cpp
 * 功能: 实现自动白平衡算法
 * 作者: WaterWhisperer
 * 创建日期: 2024-12-04
 * 更新日期: 2025-06-05
 * 版权: 自由使用，保留作者信息
 */

#include <opencv2/opencv.hpp>
#include <vector>
#include <algorithm>
#include <stdexcept>
#include <iostream>

using namespace cv;
using namespace std;

void motowhitebalance(const Mat& originalImage, Mat& adjustedImage) {
    // 将图像从RGB颜色空间转换到YCbCr颜色空间
    Mat im1;
    cvtColor(originalImage, im1, COLOR_RGB2YCrCb);
    Mat Lu, Cb, Cr;
    split(im1, vector<Mat>{Lu, Cb, Cr});
    
    // 获取图像尺寸
    int x = im1.rows;
    int y = im1.cols;

    // 定义test函数来计算平均值和差值
    auto test = [](const Mat& Cb, const Mat& Cr) {
        Scalar Mb = mean(Cb);
        Scalar Mr = mean(Cr);
        double Db = sum(Cb - Mb[0])[0] / (Cb.rows * Cb.cols);
        double Dr = sum(Cr - Mr[0])[0] / (Cr.rows * Cr.cols);
        return make_tuple(Mb[0], Mr[0], Db, Dr);
    };
    
    // 处理图像块
    Mat I1 = im1(Rect(0, 0, y / 2, x / 2));
    auto [Mb1, Mr1, Db1, Dr1] = test(Cb(Rect(0, 0, y / 2, x / 2)), Cr(Rect(0, 0, y / 2, x / 2)));

    Mat I2 = im1(Rect(0, x / 2, y / 2, x / 2));
    test(Cb(Rect(0, x / 2, y / 2, x / 2)), Cr(Rect(0, x / 2, y / 2, x / 2)));

    Mat I3 = im1(Rect(y / 2, 0, y / 2, x / 2));
    auto [Mb3, Mr3, Db3, Dr3] = test(Cb(Rect(y / 2, 0, y / 2, x / 2)), Cr(Rect(y / 2, 0, y / 2, x / 2)));

    Mat I4 = im1(Rect(y / 2, x / 2, y / 2, x / 2));
    auto [Mb4, Mr4, Db4, Dr4] = test(Cb(Rect(y / 2, x / 2, y / 2, x / 2)), Cr(Rect(y / 2, x / 2, y / 2, x / 2)));

    // 计算整体平均值和差值
    double Mr = (Mr1 + Mr3 + Mr4) / 3;
    double Mb = (Mb1 + Mb3 + Mb4) / 3;
    double Dr = (Dr1 + Dr3 + Dr4) / 3;
    double Db = (Db1 + Db3 + Db4) / 3;

    // 提取亮度值
    vector<uchar> Ciny;
    Mat tst = Mat::zeros(Lu.size(), CV_8UC1);
    for (int i = 0; i < x; i++) {
        for (int j = 0; j < y; j++) {
            double b1 = Cb.at<uchar>(i, j) - (Mb + Db * (Mb > 0 ? 1 : -1));
            double b2 = Cr.at<uchar>(i, j) - (1.5 * Mr + Dr * (Mr > 0 ? 1 : -1));
            if (abs(b1) < abs(1.5 * Db) && abs(b2) < abs(1.5 * Dr)) {
                Ciny.push_back(Lu.at<uchar>(i, j));
                tst.at<uchar>(i, j) = Lu.at<uchar>(i, j);
            }
        }
    }

    // 排序和选择
    sort(Ciny.begin(), Ciny.end(), greater<uchar>());
    int count = round(Ciny.size() / 10);
    vector<uchar> Ciny2(Ciny.begin(), Ciny.begin() + count);
    uchar mn = *min_element(Ciny2.begin(), Ciny2.end());
    Mat index = Lu > mn;

    // 计算增益
    Mat R, G, B;
    split(originalImage, vector<Mat>{B, G, R}); // OpenCV 使用 BGR 格式

    // 确保通道不为空
    if (R.empty() || G.empty() || B.empty()) {
        throw runtime_error("Error: One of the color channels is empty.");
    }

    Scalar Rave = mean(R, index);
    Scalar Gave = mean(G, index);
    Scalar Bave = mean(B, index);
    uchar Ymax = *max_element(Lu.begin<uchar>(), Lu.end<uchar>());

    double Rgain = static_cast<double>(Ymax) / Rave[0];
    double Ggain = static_cast<double>(Ymax) / Gave[0];
    double Bgain = static_cast<double>(Ymax) / Bave[0];

    // 调整图像
    R = R * Rgain;
    G = G * Ggain;
    B = B * Bgain;

    // 合并通道
    vector<Mat> channels = {B, G, R}; // OpenCV 使用 BGR 格式
    merge(channels, adjustedImage);
}

int main(int argc, char** argv) {
    // 检查命令行参数
    if (argc != 2) {
        std::cerr << "用法: " << argv[0] << " <输入图像路径>" << std::endl;
        return -1;
    }
    
    // 读取图像
    Mat originalImage = imread(argv[1]);
    if (originalImage.empty()) {
        std::cerr << "错误: 无法读取图像: " << argv[1] << std::endl;
        return -1;
    }
    
    // 应用白平衡调整
    Mat adjustedImage;
    motowhitebalance(originalImage, adjustedImage);
    
    // 显示图像
    imshow("Original Image", originalImage);
    imshow("Adjusted Image", adjustedImage);
    waitKey(0);
    destroyAllWindows();

    return 0;
}

File: tools/imageDefogging/scripts/autoWhiteBalance.py
'''
文件名：autoWhiteBalance.py
功能：实现自动白平衡算法
作者: WaterWhisperer
创建日期: 2024-12-04
更新日期: 2025-06-05
版权: 自由使用，保留作者信息
'''

import cv2
import numpy as np

def motowhitebalance(originalImage):
    """
    实现自动白平衡算法
    Args:
        originalImage: 输入的BGR格式图像
    Returns:
        adjustedImage: 白平衡调整后的图像
    """
    # 将图像从RGB颜色空间转换到YCbCr颜色空间
    # YCbCr更适合处理色彩信息
    im1 = cv2.cvtColor(originalImage, cv2.COLOR_RGB2YCrCb)
    Lu, Cb, Cr = cv2.split(im1)  # Lu为亮度分量，Cb和Cr为色度分量
    
    # 获取图像尺寸用于分块处理
    x, y, _ = im1.shape
    
    def test(Cb, Cr, x, y):
        """
        计算色度分量的统计特征
        Args:
            Cb, Cr: 色度分量
            x, y: 块的尺寸
        Returns:
            Mb, Mr: 色度分量的平均值
            Db, Dr: 色度分量的偏差
        """
        Mb = np.mean(Cb)
        Mr = np.mean(Cr)
        Db = np.sum(Cb - Mb) / (x * y)
        Dr = np.sum(Cr - Mr) / (x * y)
        return Mb, Mr, Db, Dr
    
    # 将图像分为四个块进行处理，提高局部自适应性
    # 第一块（左上）
    I1 = im1[:x//2, :y//2, :]
    Mb1, Mr1, Db1, Dr1 = test(Cb[:x//2, :y//2], Cr[:x//2, :y//2], x//2, y//2)
    
    # 第二块（左下）
    I2 = im1[x//2:, :y//2, :]
    _, _, _, _ = test(Cb[x//2:, :y//2], Cr[x//2:, :y//2], x//2, y//2)
    # Mb2, Mr2, Db2, Dr2 = test(Cb[:x//2, :y//2], Cr[:x//2, :y//2], x//2, y//2)

    # 第三块（右上）
    I3 = im1[:x//2, y//2:, :]
    Mb3, Mr3, Db3, Dr3 = test(Cb[:x//2, y//2:], Cr[:x//2, y//2:], x//2, y//2)
    
    # 第四块（右下）
    I4 = im1[x//2:, y//2:, :]
    Mb4, Mr4, Db4, Dr4 = test(Cb[x//2:, y//2:], Cr[x//2:, y//2:], x//2, y//2)
    
    # 计算整体统计特征（忽略第二块，水下图像左下区域通常容易受影响）
    # 经验表明，左下区域可能包含更多阴影或噪声，因此忽略该区域的统计特征
    Mr = (Mr1 + Mr3 + Mr4) / 3
    Mb = (Mb1 + Mb3 + Mb4) / 3
    Dr = (Dr1 + Dr3 + Dr4) / 3
    Db = (Db1 + Db3 + Db4) / 3
    
    # 如果需要考虑第二块，包含所有四个块
    # Mr = (Mr1 + Mr2 + Mr3 + Mr4) / 4
    # Mb = (Mb1 + Mb2 + Mb3 + Mb4) / 4
    # Dr = (Dr1 + Dr2 + Dr3 + Dr4) / 4
    # Db = (Db1 + Db2 + Db3 + Db4) / 4

    # 提取符合条件的亮度值
    Ciny = []
    tst = np.zeros_like(Lu)
    for i in range(x):
        for j in range(y):
            # 检查色度值是否在阈值范围内
            b1 = Cb[i,j] - (Mb + Db * np.sign(Mb))
            b2 = Cr[i,j] - (1.5 * Mr + Dr * np.sign(Mr))
            if (b1 < abs(1.5 * Db) and b2 < abs(1.5 * Dr)):
                Ciny.append(Lu[i,j])
                tst[i,j] = Lu[i,j]
    
    # 对亮度值进行排序和选择
    Ciny = np.array(Ciny)
    sumsort = np.sort(Ciny)[::-1]  # 降序排序
    count = round(len(sumsort) / 10)  # 选取前10%的亮度值
    Ciny2 = sumsort[:count]
    mn = np.min(Ciny2)
    index = Lu > mn  # 创建亮度蒙版
    
    # 计算每个通道的增益
    R = originalImage[:,:,0]
    G = originalImage[:,:,1]
    B = originalImage[:,:,2]
    Rave = np.mean(R[index])
    Gave = np.mean(G[index])
    Bave = np.mean(B[index])
    Ymax = np.max(Lu)
    
    # 计算增益系数
    Rgain = Ymax / Rave
    Ggain = Ymax / Gave
    Bgain = Ymax / Bave
    
    # 应用增益调整
    R = R * Rgain
    G = G * Ggain
    B = B * Bgain
    adjustedImage = cv2.merge((R, G, B))
    
    return adjustedImage

# 主程序
if __name__ == "__main__":
    # 读取图像
    originalImage = cv2.imread('/home/water/桌面/ROV/data1.jpg')
    if originalImage is None:
        print('Could not read the image.')
        exit(1)
    
    # 应用白平衡调整
    adjustedImage = motowhitebalance(originalImage)
    
    # 显示结果
    cv2.namedWindow('Original Image', cv2.WINDOW_NORMAL)
    cv2.imshow('Original Image', originalImage)
    cv2.namedWindow('Adjusted Image', cv2.WINDOW_NORMAL)
    cv2.imshow('Adjusted Image', adjustedImage)
    cv2.waitKey(0)
    cv2.destroyAllWindows()